common/meparser.cpp

```diff
@@ -103,88 +103,135 @@ USTATUS MeParser::parseMeRegionBody(const UModelIndex & index)
     }
-    
-    // Populate partition table header
-    const FPT_HEADER* ptHeader = (const FPT_HEADER*)region.constData();
+
     UINT32 romBypassVectorSize = 0;
     if (region.left(sizeof(UINT32)) != FPT_HEADER_SIGNATURE) {
-        // Adjust the header to skip ROM bypass vector
+        // May be v20|v21?
         romBypassVectorSize = ME_ROM_BYPASS_VECTOR_SIZE;
-        ptHeader = (const FPT_HEADER*)(region.constData() + romBypassVectorSize);
+        if (region.mid(romBypassVectorSize, sizeof(UINT32)) != FPT_HEADER_SIGNATURE) {
+            msg(usprintf("%s: ME region has no FPT header", __FUNCTION__), parent);
+            return U_INVALID_ME_PARTITION_TABLE;
+        }
     }
-    
+
+    const FPT_HEADER10* pt1Header = (const FPT_HEADER10*)(region.constData() + romBypassVectorSize);
+
     // Check region size again
-    UINT32 ptBodySize = ptHeader->NumEntries * sizeof(FPT_HEADER_ENTRY);
-    UINT32 ptSize = romBypassVectorSize + sizeof(FPT_HEADER) + ptBodySize;
+    UINT32 ptBodySize = pt1Header->NumEntries * sizeof(FPT_HEADER_ENTRY);
+    UINT32 ptSize = pt1Header->HeaderLength + ptBodySize;   // HeaderLength accounts for ROMB
     if ((UINT32)region.size() < ptSize) {
-        msg(usprintf("%s: ME region too small to fit partition table", __FUNCTION__), parent);
+        msg(usprintf("%s: ME region too small to fit FPT", __FUNCTION__), parent);
         return U_INVALID_ME_PARTITION_TABLE;
     }
-    
-    // Recalculate checksum
-    UByteArray tempHeader = UByteArray((const char*)ptHeader, sizeof(FPT_HEADER));
-    FPT_HEADER* tempPtHeader = (FPT_HEADER*)tempHeader.data();
-    tempPtHeader->Checksum = 0;
-    UINT8 calculated = calculateChecksum8((const UINT8*)tempPtHeader, sizeof(FPT_HEADER));
-    bool msgInvalidPtHeaderChecksum = (calculated != ptHeader->Checksum);
-    
+
+    UINT32 fptNumEntries;
+    UINT32 chksumCalculated;
+    bool msgInvalidPtHeaderChecksum;
+
+    // Verify checksum
+    switch (pt1Header->HeaderVersion) {
+        case 0x10:
+            /* FALLTHROUGH */
+        case 0x20:
+            {
+                UByteArray tempHeader = region.left(pt1Header->HeaderLength);
+                tempHeader[romBypassVectorSize + 0x0B] = 0x00;
+                chksumCalculated = calculateChecksum8((const UINT8 *)tempHeader.constData(), tempHeader.size());
+                msgInvalidPtHeaderChecksum = (chksumCalculated != pt1Header->Checksum);
+            }
+            fptNumEntries = pt1Header->NumEntries;
+            break;
+        case 0x21:
+            msg(usprintf("%s: FPT header version 21h checksum verification will be coded later", __FUNCTION__, pt1Header->HeaderVersion), parent);
+            return U_INVALID_ME_PARTITION_TABLE;
+        default:
+            msg(usprintf("%s: FPT header with unknown version %02h", __FUNCTION__, pt1Header->HeaderVersion), parent);
+            return U_INVALID_ME_PARTITION_TABLE;
+    }
+
+    // XXX: Shall we continue with parsing?
```

Comment:
Need to continue here, and in cases of unknown/not implemented checksums above.

Reply:
Well, just drop line 170 (return ...)

Reply:
I have now idea how it can be done! Just learning to employ that stuff (pull requests).

Reply:
Add another commit to your branch that resolves the issue, the PR will be updated automatically.